[FEATURE] Simplifier le formulaire de réconciliation SUP (PIX-2415)

bpetetot opened the pull request:
## :unicorn: Problème

Lors de la mise en place de l’import pour le SUP, nous avons remarqué que le système de réconciliation actuel permettant de se connecter sans numéro étudiant créé de nombreux doublons / triplons d'étudiants.

## :robot: Solution

Nous avons décidé de simplifier le processus et de “limiter” les entrées des étudiants si ceux-ci n'ont pas de numéro étudiant.

Nous modifions la page de réconciliation pour afficher directement le numéro étudiant, prénom, nom et date de naissance.
Il est obligatoire de saisir toutes ces informations pour valider le formulaire.

Pour être réconcilié:
- Le numéro étudiant doit être retrouvé dans les schooling registrations (à la casse près)
- Au moins une schooling registration doit avoir:
  - Le nom / prénom (une distance d'édition approximative)
  - La date de naissance exacte saisie

Si on trouve exactement une schooling registration alors on réconcilie l'utilisateur (intégration du userID au sein du schooling registration).

Si on ne trouve aucune ou plusieurs schooling registrations, alors on affiche un message d’erreur. 

## :rainbow: Remarques

- Le helper de test `fillInByLabel(label: string, value: string)` a été ajouté (copié/collé de la version de pix-orga)

## :100: Pour tester

Se connecter avec `userpix1` et accéder à la campagne `AZERTY789`

**Test 1:**
Oublier un des paramètres suivant: numéro étudiant, nom, prénom, date de naissance.
> Des messages d'erreur doivent apparaître pour chaque champs manquants.

**Test 2:**
Saisir les infos suivantes:
```
numero etudiant = XXX // invalide
prenom = Joffrey
nom = Baratheon
date de naissance = 28/02/2000
```
> Un message d'erreur doit apparaitre

**Test 3:**
Saisir les infos suivantes:
```
numero etudiant = JAIMELESFRUITS123
prenom = Magalie  // invalide
nom = Baratheon
date de naissance = 28/02/2000
```
> Un message d'erreur doit apparaitre

**Test 4:**
Saisir les infos suivantes:
```
numero etudiant = JAIMELESFRUITS123
prenom = Joffrey
nom = Lanister   // invalide
date de naissance = 28/02/2000
```
> Un message d'erreur doit apparaitre

**Test 5:**
Saisir les infos suivantes:
```
numero etudiant = JAIMELESFRUITS123
prenom = Joffrey
nom = Baratheon
date de naissance = 28/02/2001    // invalide
```
> Un message d'erreur doit apparaitre

**Test 6:**
Saisir les infos suivantes:
```
numero etudiant = JAIMELESFRUITS123
prenom = Joffrey
nom = Baratheon
date de naissance = 28/02/2000
```
> L'utilisateur doit pouvoir accéder à la campagne

**Test 7:**
Se connecter avec `jaune.attend` et accéder à la campagne `AZERTY789`
Saisir les infos suivantes:
```
numero etudiant = JAIMELESFRUITS123
prenom = Joffrey
nom = Baratheon
date de naissance = 28/02/2000
```
> Un message d'erreur doit apparaitre (car schooling registration déjà reconciliée)
```
Veuillez vérifier les informations saisies, ou si vous avez déjà un compte Pix, connectez-vous avec celui-ci.
```

bpetetot committed:
[APP] Add fillInByLabel test helper function

fillInByLabel fill a text value to an input for the given label or aria-label

bpetetot committed:
[APP] Rework SUP Join form

Now Student number, first name, last name and birth date are mandatory fields.

pix-service commented:
I'm deploying this PR to these urls:

- App: https://app-pr2893.review.pix.fr
- Orga: https://orga-pr2893.review.pix.fr
- Certif: https://certif-pr2893.review.pix.fr
- Admin: https://admin-pr2893.review.pix.fr
- API: https://api-pr2893.review.pix.fr/api/

 Please check it out!

bpetetot committed:
[API] Set student number as required when student registers

bpetetot committed:
[APP] Rename higher schooling registration reconciliation route

From POST /api/schooling-registration-user-associations/register

To POST /api/schooling-registration-user-associations/student

bpetetot committed:
[API] Rework higher schooling registration reconciliation

Reconciliation is done when a schooling registration is found for a student number (case insensitive), a birthdate, and firstname/lastname (approximative). Otherwise it throws errors

bpetetot committed:
[API] Remove saveAndReconcile function

bpetetot committed:
[API] Rework findOneRegisteredByOrganizationIdAndUserData function

Defined in higher-schooling-registration-repository

Make student number and birthdate mandatory in the query

bpetetot commented on api/lib/infrastructure/repositories/higher-schooling-registration-repository.js:
```diff
@@ -98,8 +98,8 @@ module.exports = {
       .query((qb) => {
         qb.where('organizationId', organizationId);
         qb.where('isSupernumerary', false);
```
**Question:**
Est-ce que je dois enlever `isSupernumerary=false` dans la requête, ou bien c'est un travail qui sera réalisé plus tard ? (sachant qu'il y a encore énormément de référence à `isSupernumerary` dans le code.

bpetetot replied:
A enlever et également dans les autres tests

bpetetot commented on mon-pix/app/components/routes/campaigns/restricted/join-sup.js:
```diff
@@ -51,6 +49,26 @@ export default class JoinSup extends Component {
     return [yearOfBirth, monthOfBirth, dayOfBirth].join('-');
   }
 
+  get isValidForm() {
+    this._validateInputString('studentNumber', this.studentNumber);
+    this._validateInputString('firstName', this.firstName);
+    this._validateInputString('lastName', this.lastName);
+    this._validateInputDay('dayOfBirth', this.dayOfBirth);
+    this._validateInputMonth('monthOfBirth', this.monthOfBirth);
+    this._validateInputYear('yearOfBirth', this.yearOfBirth);
+
+    const hasError = (
+      this.errors.studentNumber ||
+      this.errors.firstName ||
+      this.errors.lastName ||
+      this.errors.dayOfBirth ||
+      this.errors.monthOfBirth ||
+      this.errors.yearOfBirth
+    );
+
+    return !hasError;
+  }
```
Vérifier s'il y a bien les tests de chaque validation

bpetetot commented on mon-pix/tests/unit/components/routes/campaigns/restricted/join-sup-test.js:
```diff
@@ -110,7 +106,7 @@ describe('Unit | Component | routes/campaigns/restricted/join-sup', function() {
       // then
       sinon.assert.notCalled(onSubmitToReconcileStub);
-      expect(component.validation.studentNumber).to.equal('Votre numéro étudiant n’est pas renseigné.');
+      expect(component.errors.studentNumber).to.equal('Votre numéro étudiant n’est pas renseigné.');
```
Tester les autres champs en erreur

bpetetot commented on mon-pix/translations/fr.json:
```diff
@@ -596,10 +596,13 @@
                     "year-error": "Votre année de naissance n’est pas valide.",
-                    "year-format": "AAAA"
+                    "year-format": "AAAA",
+                    "year-label": "année de naissance (AAAA)"
```
Supprimer les JJ, MM, AAAA des labels

bpetetot commented on api/tests/acceptance/application/schooling-registration-user-association-controller_test.js:
```diff
@@ -718,14 +718,18 @@ describe('Acceptance | Controller | Schooling-registration-user-associations', (
-      return knex('schooling-registrations').delete();
-    });
-
     it('should return an 204 status after creating higher schooling registration', async () => {
```
should return an 204 status after updating higher schooling registration

lisequesnel reviewed (approved):
validé fonctionnellement en mob ✅ 
validé techniquement ✅
